### ISSUE TITLE
Tridactyl (Beta?) is very broken on some Linux distributions on FF73+

### ISSUE THREAD

**Issue by bovine3dom** [tridactyl/tridactyl#2180]
Version: 1.17.1pre3587

- New tab page doesn't show
- Command line + mode indicator don't show up
- Many excmds don't work - e.g. j/k
- but quickmarks work
- about:debugging seems to have a mostly functional `tri` object

Presumably #2171 broke it all.

**Comment by PMarci** [comment]
The same happens on Ubuntu 18.04 and version 1.17.1pre3587.

**Comment by bovine3dom** [comment]
I've reverted the changes in 27572a5. Hopefully the next beta will work.

**Comment by bovine3dom** [comment]
Closing as it now works on my machine :) Let me know if it's broken on yours.

**Comment by PMarci** [comment]
I first disabled and enabled automatic updates for `Tridactyl: Beta` in `about:addons`, this seemingly picked up the new beta build. I restarted FF, `about:addons` now shows me version `1.17.1pre3588` for the addon.
Then I installed `tridactyl-latest.xpi` from the [link in the README](https://tridactyl.cmcaine.co.uk/betas/?sort=time&order=desc) and restarted anew but I suspect this was the same artifact anyway. Nothing seems to be fixed though :cry:.

**Comment by bovine3dom** [comment]
It's definitely fixed on my machine on Windows with 3588 so I'm wondering if yours is a different issue. Could you try old betas from https://tridactyl.cmcaine.co.uk/betas and let us know which if any fixes it? (Starting with each try).

**Comment by PMarci** [comment]
Well, I went back to `1.17.1pre3571`, removing the previous version and restarting FF with each try, and no still no luck.

> I'm wondering if yours is a different issue. 

I doubt it, as all issues you mentioned correspond to mine, one addition I could mention is that hinting still works, but seemingly with a different color scheme as before. I don't think I can keep trying right now, as I'm supposed to be at work, but don't you think it could be related to the FF update from 72 to 73? Considering the couple previous versions I just tried were all working before I updated FF. In my initial post I was also referring to updating FF, not the extension.

**Comment by PMarci** [comment]
I'll try to downgrade FF to 72 and also check my Windows partition where I never installed Tridactyl later today.

**Comment by bovine3dom** [comment]
> Well, I went back to 1.17.1pre3571, removing the previous version and restarting FF with each try, and no still no luck.

This means that the changes I made to config synchronisation aren't the cause, so I think it is a different issue to the one I had, as reverting those changes has fixed my install (I am also on FF73).

It could be related to the Firefox update. Ubuntu have broken Tridactyl in the past with their tweaks to Firefox. You could try installing an unadultered build straight from Mozilla: https://help.ubuntu.com/community/FirefoxNewVersion/MozillaBuilds

Thanks for your time on debugging this :)

**Comment by PMarci** [comment]
Ok, so far I've nuked my old firefox 73 install from `/opt/` and installed 72 in `~/firefox/`, deleted `compatibility.ini` from my old profile, and started using it with this new 72 install. My previous install wasn't a Ubuntu install in the first place (in `/opt/`), but just to be safe I uninstalled every firefox-related package from my system (there was a ubuntu firefox install as well, which I didn't use and it had its own profile). I then installed `1.17.1pre3588` on 72 and am experiencing the same problem.
Now I'll try launching an instance with the newly created default profile I was forced to create and I'll try installing `1.17.1pre3588` there as well.

**Comment by PMarci** [comment]
Installing `1.17.1pre3588` on this new profile on FF 72 solves the issue for that profile. I assume it would work for 73 as well. The question remains what I'd have to purge from my profile to make it work again. I don't want to dump the entire profile, it's a big snowball that I've been rolling along for years, I don't even want to lose the history. In terms of Tridactyl, I only care about my `.tridactylrc`  and that wouldn't be impacted anyway, and I also have it backed up. I of course have backups of my FF profile as well, so I wouldn't mind exploring the possibilites.

**Comment by chrispickard** [comment]
fwiw I am on firefox nightly on centos7 and 1.17.1pre3588 shows the same symptoms

the console shows 
```
TypeError: "can't access property "document", l.contentWindow is null"
    d commandline_content.ts:60
    showcmdline .excmds_content.generated.ts:2367
    fillcmdline_notrail .excmds_content.generated.ts:2394
    u controller.ts:19
    d controller_content.ts:184
    p controller_content.ts:198
    R content.ts:51
    L content.ts:65
    <anonymous> content.ts:77
```
when I attempt to open the cmdline

**Comment by bovine3dom** [comment]
Arch, FF73 and 1.17.1pre3588 work fine for me on two separate machines so this is going to be tricky to debug.

That error @chrispickard is just a generic "the command line doesn't exist" which means that Tridactyl is very broken, so unfortunately it isn't very helpful for determining the issue. FF Nightly does break stuff with some frequency (it is sort of the point of it after all) so I wouldn't be surprised if it was a different issue to @PMarci's.

If it's a problem with the profile it could be that Tridactyl's configuration has become very mangled - in which case you should probably reset the settings at the bottom of the options page (about:addons -> Tridactyl -> Preferences, scroll all the way down). If that doesn't work we can try other things :)

**Comment by chrispickard** [comment]
hmmmm... resetting the config didn't take care of it, but it may be a nightly issue because I've attempted to use previous betas (that worked before) and they are still not working. I'll see if nightly updates fix this issue tomorrow. thanks for the help

**Comment by PMarci** [comment]
I reset the config and it didn't fix the issues, I'm at where I started in this issue. I should note that Tridactyl and Firefox (I could've sworn that I set the latter to not update itself, must've gotten lost on profile switch) updated themselves in the meantime, now I'm back at 1.17.1pre3588 on FF73.

**Comment by PMarci** [comment]
Also the console output is very similar for me:
```
TypeError: "l.contentWindow is null"
    d commandline_content.ts:60
    showcmdline .excmds_content.generated.ts:2367
    fillcmdline_notrail .excmds_content.generated.ts:2394
    u controller.ts:19
    d controller_content.ts:184
    p controller_content.ts:198
    R content.ts:51
commandline_content.ts:64:6
    d commandline_content.ts:64
    showcmdline .excmds_content.generated.ts:2367
    fillcmdline_notrail .excmds_content.generated.ts:2394
    u controller.ts:19
    d controller_content.ts:184
    p controller_content.ts:198
    R content.ts:51
```

**Comment by bovine3dom** [comment]
@PMarci does it work on FF73 on a fresh profile?

**Comment by PMarci** [comment]
@bovine3dom Yes it does.
However when selecting the profile I had to create for the new FF 72 installation (which has since updated to 73) named `default` the same issues arise. :thinking:
That is if I'm not mixing up `default` `default-release-1` and `default-release-2`, but I'm pretty sure I only ever used `default` yesterday and the rest are some automatically created fallbacks or artifacts of earlier installation. During all this testing, I never used them.

**Comment by PMarci** [comment]
Oh and to avoid any confusion I should point out again that I only ever used Mozilla builds unpacked from `tar.bz2` archives.

**Comment by bovine3dom** [comment]
@PMarci let's try forcing Tridactyl to reset itself. On a Firefox profile where Tridactyl is installed but broken:

- go to about:debugging
- click "This Firefox"
- scroll down to Tridactyl
- click "Inspect"
- click the "console" tab
- type browser.storage.sync.clear().then(x=>console.log("Sync cleared")) and press Enter
- type browser.storage.local.clear().then(x=>console.log("Local cleared")) and press Enter
- wait for "Sync cleared" and "Local cleared" to appear in the console
- Close Firefox and start it again
- Cross your fingers and hope that Tridactyl works :)

**Comment by PMarci** [comment]
@bovine3dom That worked!
Thank you very much for your help! I was starting to get frustrated for having to reach for my mouse even after a day...
Strangely, after I did your fix, FF started with the fresher-than-fresh test profile I created for my last reply, and in that profile Tridactyl was broken only after one restart and no version upgrade neither for the extension, nor for FF...
Also, this is what I saw on the debugging tab, maybe it's of some interest...
![tridebug](https://user-images.githubusercontent.com/22032937/75027869-8cfac780-549f-11ea-86f0-1e864279198e.png)
Thanks again for helping out! I'd love to contribute, sadly I don't know the first thing about browser extension development, and only had the slightest of whiffs of JS... and my schedule currently doesn't really allow getting into either of these.

**Comment by PMarci** [comment]
I noticed another issue just now, in my `.tridactylrc` I have the following to disable hinting on youtube videos:
`unbindurl youtube.com/watch --mode=normal f`
This used to work, but now it doesn't, even after calling `:source ~/.tridactylrc` or the command itself explicitly. The command line says 
`:# An error occurred in the content controller: TypeError: n is undefined`
Switching to `ignore` mode achieves the behavior I want (other than disabling everything else as well).
If this sounds like an existing issue, could you point me to it? If not I'll open one as soon as I have time.

**Comment by PMarci** [comment]
Bad news, after a restart the command line is back to being broken. :disappointed: 
Strangely though, I noticed that the status indicator flashes up for a brief moment on each tab I switch to to trigger a reload...

**Comment by bovine3dom** [comment]
Maybe something in your RC is breaking it? Could you reset the settings again via about:debugging and move the RC file elsewhere before restarting?

**Comment by PMarci** [comment]
Moved my `.tridactylrc`, restarted multiple times after the fix and the command line didn't break again.
However the command `unbindurl youtube.com/watch --mode=normal f` still has no effect, with the same error message presented when pressing `f`.
![image](https://user-images.githubusercontent.com/22032937/75033622-99851d00-54ab-11ea-9479-3c4828662520.png)
Here are the contents of my RC file for reference, nothing exotic IMO, most of it I got from some example in a gist:
```
sanitise tridactyllocal tridactylsync
" Ctrl-F should use the browser's native 'find' functionality.
unbind <C-f>

" various focus options
" doesn't work ATM
set newtabfocus page
set allowautofocus false
setpref browser.autofocus false

" hinting
set hintfiltermode vimperator-reflow
set hintchars 1234567890
set hintnames short

" GUI options
colors dark
guiset_quiet statuspanel right
set modeindicatorshowkeys true

" But also support Tridactyl search too.
bind / fillcmdline find
bind ? fillcmdline find -?
bind n findnext 1
bind N findnext -1
set incsearch true

" Ctrl C for copying in the command line
unbind --mode=ex <C-c>

" x as prev tab, unbind default
bind --mode=normal x tab #
unbind --mode=normal <C-6>
unbind --mode=normal <C-^>

" Hint only titles and menu options on yt
bindurl youtube.com/feed f hint -Jc .yt-simple-endpoint:not(.ytd-thumbnail)
bindurl youtube.com/feed F hint -Jbc .yt-simple-endpoint:not(.ytd-thumbnail)
bindurl youtube.com/channel f hint -Jc .yt-simple-endpoint:not(.ytd-thumbnail)
bindurl youtube.com/channel F hint -Jc .yt-simple-endpoint:not(.ytd-thumbnail)

" Hint only results on ddg
bindurl duckduckgo.com/ f hint -Jc .result__title .result__a
bindurl duckduckgo.com/ F hint -Jbc .result__title .result__a

" disable hinting on media sites where I need f shortcuts
unbindurl youtube.com/watch --mode=normal f
unbindurl youtube.com/watch --mode=normal F
unbindurl escapistmagazine.com/v2/20 --mode=normal f
unbindurl escapistmagazine.com/v2/20 --mode=normal F

" disable mode indicator on yt
seturl youtube.com/watch modeindicator false

" quickmarks
[some quickmarks]

" allow autofocus on graphiql and localhost
seturl [site] true
seturl [site] true
seturl localhost allowautofocus true

set searchengine duckduckgo

```
Could it be that singular guiset command?

**Comment by bovine3dom** [comment]
If you source it manually, wait a while, up to a minute, and then restart does it break Tridactyl?

This is potentially very useful, thank you :)

**Comment by PMarci** [comment]
@bovine3dom I did the following before I your most recent reply. I didn't source the RC file manually or wait for any appreciable time between the steps either, here's the comment I prepared:

I placed my RC file back to home, commented out all lines with `unbindurl` and `guiset_quiet` which seemingly had no effect (applying the fix and restarting between each try). 
At this point the only command in the RC file referencing a `youtube/watch` url is the `modeindicator false` one.
This behavior is inconsistent and hard to pin down, over a handful of resets and restarts I had the following occur (without touching the RC file inbetween):
* working commandline, `modeindicator false` apparently ignored, as it appeared on both `youtube/watch` urls and other sites
* nonfunctional commandline
* nonfunctional commandline on some tabs

Here is a scenario I noted, all of this happened in sequence:
1. reset, restart
1. load into debug tool tab
1. switch to github markdown help tab, modeindicator flashes up and then disappears on every subsequent tab switch, commandline nonfunctional everywhere
1. restart without reset
1. load into debug tool tab
1. commandline nonfunctional everywhere
1. reset, restart
1. load into youtube tab
1. commandline works everywhere, modeindicator shows up on youtube/watch even though `seturl modeindicator youtube/watch false` uncommented this entire time
1. restart without reset
1. load into youtube tab
1. commandline works everywhere, modeindicator disappears after first press of `j` (or load, both happened pretty close) as it should
1. reset, restart
1. load into non-youtube tab
1. commandline nonfunctional everywhere, modeindicator shows up while tabs are reloading after being switched to but then disappears on load
1. reset, restart
1. load into debug tool tab
1. commandline works everywhere, modeindicator shows up on youtube/watch even though `seturl modeindicator youtube/watch false` uncommented this entire time
1. restart without reset
1. load into non youtube tab
1. switch to youtube tab
1. commandline works everywhere, modeindicator disappears on load as it should

There are some tab switches I didn't note, those happened before confirming that the commandline works everywhere. In this first round I wasn't able to reproduce the situation where the command line only worked on some tabs or where the color command wasn't applied.

Next I commented out the `seturl youtube.com/watch modeindicator false` and all `bindurl` commands in the RC files as well. The following went down:

1. reset, restart
1. load into debug tool tab
1. commandline works everywhere, modeindicator doesn't show up on youtube/watch even though `seturl modeindicator youtube/watch false` is commented out this time around
1. restart without reset
1. color not applied, commandline works everywhere, modeindicator doesn't show up on youtube/watch even though `seturl modeindicator youtube/watch false` is commented out this time around
1. restart without reset
1. load into youtube tab
1. color applied, commandline only works on initial youtube tab, nonfunctional in every other tab, modeindicator doesn't show up on inital youtube tab even though `seturl modeindicator youtube/watch false` is commented out this time around
1. a reload of the initial youtube tab breaks it even there
1. restart without reset
1. load into youtube tab
1. commandline nonfunctional everywhere
1. reset without restarting out of curiosity
1. switch to youtube tab
1. commandline works, tridactylrc not applied as expected
1. restart without reset (there was a reset three steps before this)
1. load into debug tool tab
1. switch to youtube tab
1. commandline works everywhere, modeindicator shows up and stays on after a couple of `j` and `k` keypresses, but curiously disappears after mousing over it (I noticed afterwards it's supposed to disappear and come back everywhere) and stays gone. Remember, during all these steps `seturl modeindicator youtube/watch false` was commented out

**Comment by bovine3dom** [comment]
I suspect a lot of the inconsistency is due to #1409 - could you try hiding your RC file, resetting Tridactyl and restarting, and just doing `:seturl youtube.com/watch modeindicator false` in Tridactyl, then restarting?

**Comment by PMarci** [comment]
@bovine3dom 
Sure, I did it and it works as expected.
Do you think it's possible that some things are applied even if they are preceded by `" ` at the start of the line?
Also this doesn't bring us much closer to finding out what (combination) breaks the command line, suggest some isolated commands that you suspect might be involved anytime. I'll apply them individually or in fragments of my original RC file that isolates a set of commands instead of commenting them out.

**Comment by bovine3dom** [comment]
https://github.com/tridactyl/tridactyl/blob/fe39c3e43adfe9cef9bc225c9db16aba3586af7c/src/background/config_rc.ts#L59 - it should be literally impossible for a line that starts with `"`to be executed - provided you haven't used a strange quote mark that looks like " but isn't " and also that it starts the line (i.e. `    " comment` would be run). OTOH it shouldn't matter even if it did run as `  " set blah` will just be an invalid ex-command which would be caught and ignored.

So Tridactyl only gets broken when it automatically sources your RC file? Does moving your RC file fix broken Tridactyl (after a restart)? What about manually sourcing the RC file?

For a broken Tridactyl,  the about:debugging console, could you run

```
browser.storage.local.get().then(x=>console.log(x))
browser.storage.sync.get().then(x=>console.log(x))
```

and post the results here (in full, after redacting any sensitive information) - you might need to use e.g. hastebin.com.

It would also be useful if you could do the same for a not-broken Tridactyl :)

**Comment by glacambre** [comment]
I think this issue has been fixed by https://github.com/tridactyl/tridactyl/pull/2202 .

**Comment by PMarci** [comment]
Sorry for disappearing, I was busy with other stuff and forgot about this issue because I left my browser and extension at a previous working version and ignored updates for a while. I can confirm the issue is gone now.
